### DISCUSSION TITLE
Unable to play video on windows.

### DISCUSSION THREAD

**Discussion by Aryan-29** [fltk-rs/fltk-rs#603] [Q&A]
Hey there.
I have added a video as an intro in my project. When I launch my program for the first time the video doesn't run properly instead a black screen appears. But when I run the code again then the video plays properly. Any idea what could be the problem?
I have added the screenshot below for your reference.
![Screenshot (50)](https://user-images.githubusercontent.com/61017218/114132064-4e350480-9921-11eb-8759-09f63fa53f3d.png)

**Comment by MoAlyousef** [comment]
Hi
Unfortunately without a minimal reproducible example I can’t tell where the problem lies. 
what other libs are you using, fltk-rs version, any logging by the video library, did you try to debug?

**Reply by Aryan-29** [reply]
fltk = "0.16.3"
vlc-rs = "0.3.0"

source code:
 
    // INTRO //
    let intro_app = App::default();
    let mut win = Window::default().with_size(500,500).center_screen();
    win.make_modal(true);
    // Create inner window to act as embedded media player
    let vlc_win = Window::new(0, 0, 500, 500, "");
    vlc_win.end();
    win.end();
    win.set_border(false);
    win.show();
    let instance = Instance::new().unwrap();
    let md = Media::new_path(&instance, "video1.mp4").unwrap();
    let mdp = MediaPlayer::new(&instance).unwrap();
    mdp.set_media(&md);
    let handle = vlc_win.raw_handle();
    // For Linux
    #[cfg(target_os = "linux")]
    mdp.set_xwindow(handle as u32);
    // For Windows
    #[cfg(target_os = "windows")]
    mdp.set_hwnd(handle);
    // For MacOS
    #[cfg(target_os = "macos")]
    mdp.set_nsobject(handle);
    mdp.set_key_input(false);
    mdp.set_mouse_input(false);
    mdp.play().unwrap();
    thread::sleep(time::Duration::from_secs(5));
    mdp.stop();
    intro_app.quit();

after playing the video I want to close the window and that is why I have used thread::sleep() method.

logs for vlc:
[000001de15754150] mmdevice audio output error: cannot initialize COM (error 0x80010106)
TagLib: MP4: No audio tracks
[000001de15774b20] avcodec decoder: Using D3D11VA (NVIDIA GeForce RTX 2070 with Max-Q Design, vendor 10de(NVIDIA), device 1f50, revision a1) for hardware decoding
[000001de17a0e730] direct3d11 vout display error: SetThumbNailClip failed: 0x800706f4

**Reply by MoAlyousef** [reply]
I'm not really sure what the issue is. Does the program in the [demos](https://github.com/MoAlyousef/fltk-rs-demos/blob/master/libvlc/src/main.rs) run correctly? (you'll have to modify the video's file name to yours)

Can you also try launching the program directly and not thru `cargo run`. I'm also not sure if the logs hint to the problem causing this. If you're using WSL on Windows, the missing audio support can cause video problems, in which case you'd have to pass the --no-audio flag to libvlc via the command-line, which can be passed like:
```rust
    let args: Vec<String> = std::env::args().collect();
    let instance = Instance::with_args(Some(args)).unwrap();
```

**Reply by Aryan-29** [reply]
Is there anything else in fltk other than vlc that can be used to play a video as I have been encountered with a lot of errors recently?

**Reply by MoAlyousef** [reply]
Yeah basically anything which accepts a raw handle, like for example gstreamer. Also ffmpeg which can translate frames to rgb images which can be drawn into a frame or opengl window (for hardware acceleration).

**Reply by Aryan-29** [reply]
gstreamer is similar to libvlc. When I package my application to be run on other machines, I get prompted with the msys2 not installed error. It runs fine on my machine because I have msys2 installed. Can you help me with ffmpeg of how to write the code for a simple video player as I am not very familiar to it.

**Reply by MoAlyousef** [reply]
It seems it's a deployment issue. You would want to link to the mingw form of the package, which wouldn't require the msys2 dlls to be installed. When you search for packages on msys2, grab the package prefixed with `mingw-w64-x86_64` or `mingw-w64-xi686` depending on the output of `echo $MINGW_PACKAGE_PREFIX`. You can even search for packages using `pacman -Ss $MINGW_PACKAGE_PREFIX-gstreamer` for example. 
Then depending on the libs installed, if the libs are shared, you would need to deploy them with your application. Otherwise, for static libs (ending with .a), you don't need to deploy with your application.
